release-operator/src/registry.rs

```diff
@@ -0,0 +1,157 @@
+use std::path::PathBuf;
+use std::str::FromStr;
+
+pub struct Registry {
+    token: SecStr,
+    crates: Vec<Crate>,
+    dry_run: bool,
+}
+
+#[derive(Clone, Debug)]
+pub struct Crate {
+    path: PathBuf,
+}
+
+enum CrateState {
+    Published,
+    Unknown,
+    Outdated,
+}
+
+impl Registry {
+    pub fn new(token: &SecStr, crates: &[Crate], dry_run: bool) -> Self {
+        Self {
+            token: token.to_owned(),
+            crates: crates.to_vec(),
+            dry_run,
+        }
+    }
+
+    pub fn publish_crates(&self) -> anyhow::Result<()> {
+        for c in &self.crates {
+            c.validate()?;
+
+            match c.determine_state()? {
+                CrateState::Published => continue,
+                CrateState::Unknown | CrateState::Outdated => {
+                    c.submit(&self.token, self.dry_run)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Crate {
+    fn validate(&self) -> anyhow::Result<()> {
+        match self.path.exists() {
+            true => Ok(()),
+            false => Err(anyhow!(
+                "given path to the '{self}' crate is either not readable or does not exist"
+            )),
+        }
+    }
+
+    fn determine_state(&self) -> anyhow::Result<CrateState> {
+        let theirs = {
+            let name = format!("{self}");
+            let search_result = cmd_lib::run_fun!(cargo search "${name}")
+                .context("search for crate on crates.io")?;
+
+            if search_result.is_empty() {
+                log::info!("{self} has not been published yet");
+                return Ok(CrateState::Unknown);
+            }
+
+            let version = cmd_lib::run_fun!(cargo search "${name}" | head -n1 | awk r#"{print $3}"# | tr -d '"')
+                .context("search crates.io for published crate version")?;
+            log::debug!("{self} found as {version} on their side");
+
+            version
+        };
+
+        let ours = {
+            let name = format!("{self}");
+            let cargo_toml_location = std::fs::canonicalize(&self.path)
+                .context("absolute path to Cargo.toml")?;
+            let mut cmd = cargo_metadata::MetadataCommand::new();
+            cmd.manifest_path(format!(
+                "{}/Cargo.toml",
+                cargo_toml_location.to_string_lossy()
+            ))
+            .no_deps();
+
+            let metadata = cmd.exec()?;
+            let package = metadata
+                .packages
+                .iter()
+                .find(|p| p.name == name)
+                .ok_or_else(|| anyhow!("could not find package"))?;
+
+            let version = package.version.to_string();
+            log::debug!("{self} found as {version} on our side");
+
+            version
+        };
+
+        if ours == theirs {
+            log::info!("{self} has already been published as {ours}");
+            return Ok(CrateState::Published);
+        }
+
+        Ok(CrateState::Outdated)
```

Comment:
I don't think this handles the case of a newer version being published on crates.io, which might happen if this is run on an old commit. Not sure if this will ever happen, and if it happens it will probably due no harm. Publishing would just fail, unless it's a commit from some alternate branch that was never published.

There's probably no need to do anything here. I'm just musing because I noticed this. We *could* detect this case and return `CrateState::SomethingIsWrongPleaseAbort` here, to be safe.

Reply:
Yep, good catch. We should handle it. Comparing versions is easy enough thanks to the `SemVer` crate, so why not.